### ISSUE TITLE
Fix gradient accumulation in trainings recipes

### ISSUE THREAD

**Issue by Optimox** [pytorch/torchtune#1878]
Hello,

A generic bug in the gradient accumulation when training LLMs has been discussed recently by unsloth.ai: https://unsloth.ai/blog/gradient 

If I understood correctly the fix would be fairly simple and would affect this line of code (this is true for all recipes):
https://github.com/pytorch/torchtune/blob/ca37c590d0a22992ec2e0236422db242383ad61b/recipes/full_finetune_single_device.py#L628

It would be something similar to this: `loss = loss / self._gradient_accumulation_steps` by `loss = loss / batch["tokens"].numel()` to avoid small samples to have more weights than large samples when accumulating.

It might not be as straightforward and a closer look should be taken here: https://github.com/unslothai/unsloth/commit/e210840ba97a5837ca59fc73ede47b56f10772dd

**Comment by ebsmothers** [comment]
Hi @Optimox thanks for creating this issue. This is on our radar already, see #1875. I need to get back to do it to do some cleanup, but hoping to land in the next couple of days here.

**Comment by Optimox** [comment]
Oh great, sorry for the duplicate!
